spec/mem/tree_borrows/memory.md

```diff
@@ -0,0 +1,298 @@
+        };
+
+        let mut allocation = self.allocations[alloc_id.0];
+
+        // Create the new child node
+        let child_node = Node {
+            children: List::new(),
+            location_states: Self::init_location_states(permission, allocation.size()),
+        };
+
+        // Add the new node to the tree
+        let child_path = allocation.extra.root.add_node(parent_path, child_node)?;
+
+        // Perform child read to all nodes
+        allocation.extra.root.access(child_path, AccessKind::Read, offset, pointee_size, true)?;
+        self.allocations.set(alloc_id.0, allocation);
+
+        // Create the child pointer and return it 
+        ret(Pointer {
+            provenance: Some((child_path, alloc_id)),
+            ..ptr
+        })
+    }
+}
+```
+
+# Memory Operations
+Then we implement the memory model interface for the Tree Borrow.
+
+### Allocate and Deallocate
+
+We create a new tree for one allocation
+
+```rust
+impl<T: Target> Memory for TreeBorrowsMemory<T> {
+    fn allocate(&mut self, kind: AllocationKind, size: Size, align: Align) -> NdResult<Pointer<Self::Provenance>>  {
+        let addr = Allocation::pick_base_address::<T>(self.allocations, size, align)?;
+
+        // Calculate the proverance for the root node
+        let alloc_id = AllocId(self.allocations.len());
+
+        // Create the root node for the tree.
+        // Initially, we set the permission as `Active`
+        let root = Node {
+            children: List::new(),
+            location_states: Self::init_location_states(Permission::Active, size),
+        };
+
+        // Path to the root node
+        let path = List::new();
+
+        let allocation = Allocation {
+            addr,
+            align,
+            kind,
+            live: true,
+            data: list![AbstractByte::Uninit; size.bytes()],
+            extra: TreeBorrowsAllocationExtra { root },
+        };
+
+        self.allocations.push(allocation);
+
+        ret(Pointer { addr, provenance: Some((path, alloc_id)) })
+    }
+}
+
+impl<T: Target> Memory for TreeBorrowsMemory<T> {
+    fn deallocate(&mut self, ptr: Pointer<Self::Provenance>, kind: AllocationKind, size: Size, align: Align) -> Result {
+        let Some((path, alloc_id)) = ptr.provenance else {
+            throw_ub!("deallocating invalid pointer")
+        };
+        // This lookup will definitely work, since AllocId cannot be faked.
+        let mut allocation = self.allocations[alloc_id.0];
+
+        allocation.deallocation_check(ptr.addr, kind, size, align)?;
+
+        // Check that ptr has the permission to write the entire allocation
+        allocation.extra.root.access(path, AccessKind::Write, Size::ZERO, size, true)?;
+
+        // Mark it as dead. That's it.
+        self.allocations.mutate_at(alloc_id.0, |allocation| {
+            allocation.live = false;
+        });
+
+        ret(())
+    }
+}
+```
+
+### Load Operation
+
+Corresponding to the `AccessKind::Read`.
```

Comment:
I don't understand the point of this comment -- AccessKind::Read is an internal Tree Borrows thing, so not sure why it would be called out here?

I think you can have alloc, dealloc, load, store in one block and just say above that this is defining the core memory operations.